Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -232,24 +232,172 @@ function Get-ServiceFabricApplicationAction
     return Get-ServiceFabricApplication @getApplicationParams
 }
 
+function Get-ServiceFabricApplicationUpgradeAction
+{
+    Param (
+        [string]
+        $ApplicationName
+    )
+
+    $global:operationId = $SF_Operations.GetApplicationUpgradeStatus
+    return Get-ServiceFabricApplicationUpgrade -ApplicationName $ApplicationName
+}
+
+function Copy-ServiceFabricApplicationPackageAction
+{
+    Param (
+        [hashtable]
+        $CopyParameters
+    )
+
+    $global:operationId = $SF_Operations.CopyApplicationPackage
+    $copyAction = { Copy-ServiceFabricApplicationPackage @CopyParameters }
+
+    Invoke-CommandWithRetries -Action $copyAction `
+        -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingCopyApplicationPackage)
+}
+
+function Register-ServiceFabricApplicationTypeAction
+{
+    Param (
+        [hashtable]
+        $RegisterParameters,
+
+        [string]
+        $ApplicationTypeName,
+
+        [string]
+        $ApplicationTypeVersion
+    )
+
+    $global:operationId = $SF_Operations.RegisterApplicationType
+    $registerAction = { Register-ServiceFabricApplicationType @RegisterParameters }
+
+    $onException = {
+        param($ex)
+        $appType = Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+        # If provisioning not started, retry register
+        if (!$appType)
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeProvisioningNotStarted)
+            return $true
+        }
+
+        # if provisioning started, wait for it to complete
+        if (($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -or ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning))
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeProvisioningStarted)
+            $appType = Wait-ServiceFabricApplicationTypeStatusChange -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+        }
+
+        # if app type got unprovisioned, retry register
+        if(!$appType)
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioned)
+            return $true
+        }
+
+        # if app type is provisioned, bail out
+        if ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available)
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeProvisioned)
+            return $false
+        }
+
+        # if provisioning failed, throw and don't retry
+        throw (Get-VstsLocString -Key SFSDK_RegisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
```

Comment:
Better to log this and return false

Reply:
OnException has three possible outcomes:
1. true - in this case retry is attempted
2. false - in this case it comes out of retry, but without failure. e.g. if get-applicationtype returns that application type is provisioned, we should come out of register-applicationtype retry loop
3. throw - it comes out of retry with failure .e.g if get-applicationtype returns that application provisioning failed